### PR TITLE
Remove plugins feature

### PR DESCRIPTION
**Pull request by canstudios-nicolaw**
Adds the ability to remove plugins. See #928 for more info.

Can delete core and uploaded components, extensions, themes and menus. Note that deleted core plugins will be re-installed when the framework is updated.

For all plugin types, the code will attempt to delete from the following places:
1) /plugins/content/[component/extension/theme/menu]/versions/[plugin]/
2) /temp/[tenant id]/adapt_framework/src/[components/extensions/theme/menu]/[plugin]/
3) The relevant entry in the xxxxtypes table
4) /plugins/content/bower/bowercache/[plugin]

(Usually the 1-3 are present for uploaded plugins and 3-4 are present for core plugins). 

Currently making this as a draft PR as it is pending final approval from QA here.

### CODE REVIEW

**Comment by canstudios-nicolaw** [comment]
Converting this to a full PR as QA have finished testing that everything gets deleted from the right places for core and non-core plugins. I am going to add tooltips to the icon buttons later this week but the rest of the feature should be good to test.

File: plugins/content/bower/index.js [hunk 1]
```diff
@@ -220,6 +351,15 @@ function initialize () {
   });
 }
 
+/**
+ * The default for listing the courses that use a specific plugin
+ * This should be overridden in each plugin type object
+ *
+ */
+BowerPlugin.prototype.getUses = function (callback, id) {
+    return callback(new Error('The getUses function must be overriden in the plugin\'s obejct'));
```
Comment (tomgreenfield) [inline plugins/content/bower/index.js]:
````suggestion
    return callback(new Error('The getUses function must be overridden in the plugin’s object'));
```

**Comment by canstudios-nicolaw** [comment]
@tomgreenfield thanks for your help, this is now ready for re-review

File: frontend/src/modules/pluginManagement/views/pluginTypeView.js [hunk 2]
```diff
@@ -38,17 +45,24 @@ define(function(require){
...
 
-      $btn.html(Origin.l10n.t('app.checking'));
+      $btn.attr({title: Origin.l10n.t('app.checking')});
+      $icon.addClass('fa-spin');
 
       $.get(this.model.urlRoot + '/checkversion/' + this.model.get('_id'), function(data) {
         if(!data.isUpdateable) {
-          $btn.attr('disabled', true).html(Origin.l10n.t('app.uptodate'));
+          $btn.attr({
+              'disabled': true,
```
Comment (tomgreenfield) [inline frontend/src/modules/pluginManagement/views/pluginTypeView.js]: Suggest removing quotes around keys from all instances of `$btn.attr({})`.

File: frontend/src/modules/pluginManagement/views/pluginTypeView.js [hunk 3]
```diff
@@ -57,23 +71,73 @@ define(function(require){
...
+        const popup = {};
+
+        if (data.courses.length === 0) {
+          Origin.Notify.confirm({
+            type: 'warning',
+            title: Origin.l10n.t('app.deleteplugin'),
+            text: Origin.l10n.t('app.confirmdeleteplugin', { plugin: this.model.get('displayName') }),
+            destructive: false,
+            callback: this.deletePluginConfirm.bind(this)
+          });
```
Comment (tomgreenfield) [inline frontend/src/modules/pluginManagement/views/pluginTypeView.js]: Suggest a `return;` – the else block is rather long.

File: frontend/src/modules/pluginManagement/views/pluginTypeView.js [hunk 4]
```diff
@@ -57,23 +71,73 @@ define(function(require){
...
+        } else {
+          var courses = '';
+          for (var i = 0, len = data.courses.length; i < len; i++) {
+            courses += data.courses[i].title + ' ' + Origin.l10n.t('app.by') + ' ' + data.courses[i].createdByEmail + '<br />'
+          }
+          popup.type = 'error';
+          popup.title = Origin.l10n.t('app.cannotdelete') + ' ' + this.model.get('displayName');
+          popup.text = '';
+          if (courses !== '') {
```
Comment (tomgreenfield) [inline frontend/src/modules/pluginManagement/views/pluginTypeView.js]: Bit confused by the logic; could `courses` ever not be populated at this point?

**Pull request merged**